plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/bootstrap/FlightClientManager.java

```diff
@@ -0,0 +1,265 @@
+        if (!isValidNode(node)) {
+            logger.warn(
+                "Unable to build FlightClient for node [{}] with role [{}] on version [{}]",
+                node.getId(),
+                node.getRoles(),
+                node.getVersion()
+            );
+            return;
+        }
+        FlightClient flightClient = buildClient(location);
+        flightClients.put(node.getId(), new ClientHolder(location, flightClient));
+    }
+
+    private void requestNodeLocation(String nodeId, CompletableFuture<Location> future) {
+        NodesFlightInfoRequest request = new NodesFlightInfoRequest(nodeId);
+        client.execute(NodesFlightInfoAction.INSTANCE, request, new ActionListener<>() {
+            @Override
+            public void onResponse(NodesFlightInfoResponse response) {
+                NodeFlightInfo nodeInfo = response.getNodesMap().get(nodeId);
+                if (nodeInfo != null) {
+                    TransportAddress publishAddress = nodeInfo.getBoundAddress().publishAddress();
+                    String address = publishAddress.getAddress();
+                    int flightPort = publishAddress.address().getPort();
+                    Location location = clientConfig.sslContextProvider != null
+                        ? Location.forGrpcTls(address, flightPort)
+                        : Location.forGrpcInsecure(address, flightPort);
+
+                    future.complete(location);
+                } else {
+                    future.completeExceptionally(new IllegalStateException("No Flight info received for node: [" + nodeId + "]"));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                future.completeExceptionally(e);
+                logger.error("Failed to get Flight server info for node: [{}] {}", nodeId, e);
+            }
+        });
+    }
+
+    private FlightClient buildClient(Location location) {
+        return OSFlightClient.builder()
+            .allocator(clientConfig.allocator)
+            .location(location)
+            .channelType(ServerConfig.clientChannelType())
+            .eventLoopGroup(clientConfig.workerELG)
+            .sslContext(clientConfig.sslContextProvider != null ? clientConfig.sslContextProvider.getClientSslContext() : null)
+            .executor(clientConfig.grpcExecutor)
+            .build();
+    }
+
+    private DiscoveryNode getNodeFromClusterState(String nodeId) {
+        return Objects.requireNonNull(clientConfig.clusterService).state().nodes().get(nodeId);
+    }
+
+    /**
+     * Closes the FlightClientManager and all associated Flight clients.
+     */
+    @Override
+    public void close() throws Exception {
+        for (ClientHolder clientHolder : flightClients.values()) {
+            clientHolder.flightClient.close();
+        }
+        flightClients.clear();
+        grpcExecutor.shutdown();
+        grpcExecutor.awaitTermination(5, TimeUnit.SECONDS);
+    }
```

Comment:
Please `clusterService.removeListener(this);`

Reply:
good point, done!